[Tang-Moyan] iP

Tang-Moyan opened the pull request:
### Duke-Rio

> “Your mind is for having ideas, not holding them.” – David Allen 

Duke-Rio helps you store your tasks in a digital list. It's,
- text-based
- automaticly saving your changes
- retrieving your previous list even if you close it by accident.

All you need to do is,

1. download it from [this link](https://github.com/Tang-Moyan/ip/releases/tag/A-Jar-release)
2. run the ip.jar
3. add your tasks.
4. let it manage your tasks for you :grin:

And it is **FREE**!

Features:

- [ ] managing tasks
- [ ] find if you have a certain task
- [ ] Managing deadlines (coming soon)

Here's the main method:
```java
public class Duke {
    public static void main(String[] args) {

            new Duke("list.txt", "data").run();


    }
}
```

You can modify this line `new Duke("list.txt", "data").run();` in the _main_ function to decide where and what file name is used to store your list.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Tang-Moyan committed:
Level 0 Rename, Greet, Exit

Tang-Moyan committed:
Added Level 1. Echo

Tang-Moyan committed:
Added Level 2. Add, List

Tang-Moyan committed:
Added Level 3. Mark as Done

Tang-Moyan committed:
Added Level 4. ToDos, Events, Deadlines

Tang-Moyan committed:
Added UI Testing

Tang-Moyan committed:
Level 5. Handle Errors

Tang-Moyan committed:
Level 6. Delete

Tang-Moyan committed:
Added Level 7. Save

Tang-Moyan committed:
Small changes

Tang-Moyan committed:
Merge branch 'branch-Level-7'

Tang-Moyan committed:
Added Leve 8. Dates and Time

Tang-Moyan committed:
Merge branch 'branch-Level-8'

AustinHuang1203 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+import java.nio.file.Path;
+import HelperClass.Task;
+
 public class Duke {
```
Hi, not sure which level you are at but you could consider more encapsulation of code 😸

Tang-Moyan replied:
Thank you very much for your review! I am a bit falling behind in catching up with the levels. I will try to get them done sooner.

AustinHuang1203 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+        printOneLine();
+    }
+
+    public  static void Exit() {
```
double spacing can change to single between public and static

AustinHuang1203 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+                System.out.println("Directory '" + directoryName + "' created.");
+            } else {
+                System.err.println("Failed to create directory '" + directoryName + "'.");
+                return;
```
redundant return statement

AustinHuang1203 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,373 @@
+
+            printOneLine();
+            switch (userInput) {
+
```
This block of code is kindda big so maybe can encapsulate out

AustinHuang1203 commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,3 @@
+todo
```
Can try to be a bit more comprehensive

AustinHuang1203 commented on src/main/java/HelperClass/Task.java:
```diff
@@ -0,0 +1,122 @@
+public class Task {
+    private boolean isDone;
+
+    private int type;
```
Can consider a more clear name beside type for int

AustinHuang1203 reviewed (commented):
Overall ok and neat code that is readable. Could consider more encapsulation and java doc comments also! But I think it is fine and nice.

Tang-Moyan committed:
partial OOP

Tang-Moyan committed:
Added A-MoreOOP

Tang-Moyan committed:
Added A-Packages (already done in previous commit)

Tang-Moyan committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-branches

Tang-Moyan committed:
Fixed the bug that cannot get user command correctly

Tang-Moyan committed:
Added A-JUnit

Tang-Moyan committed:
Created A-Jar

Tang-Moyan committed:
Added JavaDoc

Tang-Moyan committed:
Modified Coding Standard

Tang-Moyan committed:
Added Level 9. Find

Tang-Moyan committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

Tang-Moyan committed:
Resolved conflict

Tang-Moyan committed:
small modification

SimWPEric commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,179 @@
+                        1, "Null", "Null", false)));
+                userListHaveChanges = true;
+                break;
+
+            case "deadline":
+                //deadline read book /by 2022-01-01
+                ui.Speak(tasks.addTask(new Task(parser.getTaskName(),
+                        2, "Null", parser.getFirstEnteredTime(), false)));
+                userListHaveChanges = true;
+                break;
+
+            case "event":
+                //event read book /from 2022-01-01 /to 2022-01-02
+                ui.Speak(tasks.addTask(new Task(parser.getTaskName(),
+                        3, parser.getFirstEnteredTime(), parser.getSecondEnteredTime(), false)));
+                userListHaveChanges = true;
+                break;
+
+            case "delete":
+                //delete 1
+                try {
+                    int i = Integer.parseInt(parser.getTaskName()) - 1;
+                    ui.Speak(tasks.deleteTask(i));
+                    userListHaveChanges = true;
+                } catch (NumberFormatException e) {
+                    ui.Speak("need to provide an integer index of task.");
+                }
+
+                break;
+
+            case "find":
+                ui.Speak(tasks.findTask(parser.getTaskName()));
+                break;
+
+
+            default:
+                ui.Speak("OOPS!!! I'm sorry, but I don't know what that means :-(");
+
+
+            }
```
According to [Java Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html): The switch statement should have the following form: Note there is no indentation for case clauses.

SimWPEric commented on src/main/java/HelperClass/Ui.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Ui {
+
+    private String MyName;
```
Try using camelCase for variable names !

SimWPEric commented on src/main/java/HelperClass/Ui.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    /**
+     * print message in a specific format
+     * @param message the original message to be printed
```
Can be good to include an empty line between the `description` and `@param` !

SimWPEric commented on src/test/java/HelperClass/ParserTest.java:
```diff
@@ -0,0 +1,19 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class ParserTest {
+    @Test
+    public void testGetUserCommand() {
+        Parser parser = new Parser();
+        parser.processUserCommand("todo Read Book");
+        assertEquals("todo", parser.getCommand());
+    }
+
+    @Test
+    public void testGetTaskName() {
+        Parser parser = new Parser();
+        parser.processUserCommand("todo Read Book");
+        assertEquals("Read Book", parser.getTaskName());
+    }
```
According to [CS2103T Website W3.7d](https://nus-cs2103-ay2324s1.github.io/website/schedule/week3/topics.html#w3-7-unit-testing) code normally use camelCase for method names e.g., `testStringConversion` but when writing test methods, sometimes another convention is used:`unitBeingTested_descriptionOfTestInputs_expectedOutcome`.

SimWPEric commented on src/main/java/HelperClass/Task.java:
```diff
@@ -0,0 +1,129 @@
+     * return the string representing how the Task object should be stored
+     * @return string representation of the Task object
+     */
+    public String ForRecordingInTextFile() {
```
Try using camelCase for method names !

SimWPEric reviewed (commented):
Good job on your progress thus far ! **A-CheckStyle** increment will come in really handy for catching styling errors !

Tang-Moyan committed:
Level 10. GUI

Tang-Moyan committed:
Make getResponse() return correct String

getResponse() only return the user input.

Let's update the getResponse() function so it work correctly with duke
and returns the right message.

In this way, duke GUI can do its job correctly.

Tang-Moyan committed:
Add showTaskStatics() function

When we have a lot of tasks in the list, it would be convenient to
have a function that shows the user how many each type of task
there are.

Let's have a new user command, "stats", to invoke the
showTaskStatics() function.

It is a convenient way of showing task information to the user.

Tang-Moyan committed:
Add A-Assertions

We need to check some crucial assumptions at certain time during
execution of the program.

Let's use assert feature to document important assumptions that
should hold at various points in the code.

In this way, users will know more about where went wrong when using
this product.

Tang-Moyan committed:
Improved code quality

Tang-Moyan committed:
Improved code quality further

Tang-Moyan committed:
Merge pull request #2 from Tang-Moyan/branch-A-Assertions

Add A-Assertions

Tang-Moyan committed:
Merge branch 'branch-A-CodeQuality'

Tang-Moyan committed:
Add sort function

Tang-Moyan committed:
Add User Guide

Tang-Moyan committed:
Add Ui.png

Tang-Moyan committed:
More improvement to the code

Tang-Moyan committed:
More improvement to the code

Tang-Moyan committed:
More improvement to the code

Tang-Moyan committed:
Apply SLAP to shorten methods

There are several methods that are too long.

Long methods are harder for other developers to understand.

Shortening them improves readability of the code.

Let's divide long methods into short methods so that other
developers can understand the code more easily.

Tang-Moyan committed:
Limit task names to be non-empty

Task names can be empty based on user input.

Empty task name is should not be allowed becuase it will lead to
bugs.

Limiting the task name to be non-empty prevents bugs.

Let's limit our user to give only non-empty task names, so that no
bugs will occur.

Tang-Moyan committed:
Command classes: extract command functions into classes

Command functions (todo, deadline, event...) have some
common behaviors (return a message).

Common behaviors should be abstract out for better OOP practice.

Abstracting out common behaviors into a super class and making each
command function a separate class allow more conveninent extensions
of current commands in the future.

Let's have a super class Command and define all commands as child
classes of the Command class.

Using inheritance is preferable over composition in this situation
because the common behaviors are not composable.
